- Description: dasd: I/O errors during DASD device qualification.
  Symptom:     I/O errors while vary on/off channel path.
  Problem:     If CIO returns -EIO for a ccw_device_start (e.g. because of
               channel path re-detection), the current I/O is set to failed.
               This leads to an immediate I/O error.
  Solution:    Keep the request in status QUEUED to enable retries and use
               higher default retry counter.
  Archs:       s390, s390x
  Problem-ID:  --
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/block/dasd.c	20 Apr 2004 13:49:52 -0000	1.137.2.3
+++ linux-2.5/drivers/s390/block/dasd.c	10 May 2004 12:46:18 -0000	1.137.2.4
@@ -7,7 +7,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999-2001
  *
- * $Revision: 1.137.2.3 $
+ * $Revision: 1.137.2.4 $
  */
 
 #include <linux/config.h>
@@ -751,16 +751,16 @@
 		break;
 	case -EBUSY:
 		DBF_DEV_EVENT(DBF_ERR, device, "%s",
-			      "device busy, retry later");
+			      "start_IO: device busy, retry later");
 		break;
 	case -ETIMEDOUT:
 		DBF_DEV_EVENT(DBF_ERR, device, "%s",
-			      "request timeout - terminated");
+			      "start_IO: request timeout, retry later");
+		break;
 	case -ENODEV:
 	case -EIO:
-		cqr->status = DASD_CQR_FAILED;
-		cqr->stopclk = cqr->startclk;
-		dasd_schedule_bh(device);
+		DBF_DEV_EVENT(DBF_ERR, device, "%s",
+			      "start_IO: device gone, retry");
 		break;
 	default:
 		DEV_MESSAGE(KERN_ERR, device,
@@ -1008,8 +1008,9 @@
 				if (device->discipline->start_IO(next) == 0)
 					expires = next->expires;
 				else
-					MESSAGE(KERN_WARNING, "%s",
-						"Interrupt fastpath failed!");
+					DEV_MESSAGE(KERN_DEBUG, device, "%s",
+						    "Interrupt fastpath "
+						    "failed!");
 			}
 		}
 	} else {		/* error */
@@ -1018,8 +1019,8 @@
 		if (cqr->dstat)
 			memcpy(cqr->dstat, irb, sizeof (struct irb));
 		else
-			MESSAGE(KERN_ERR, "%s",
-				"no memory for dstat...ignoring");
+			DEV_MESSAGE(KERN_ERR, device, "%s",
+				    "no memory for dstat...ignoring");
 #ifdef ERP_DEBUG
 		/* dump sense data */
 		dasd_log_sense(cqr, irb);

--- linux-2.5/drivers/s390/block/dasd_eckd.c	20 Apr 2004 13:28:03 -0000	1.53.2.1
+++ linux-2.5/drivers/s390/block/dasd_eckd.c	10 May 2004 12:46:18 -0000	1.53.2.2
@@ -7,7 +7,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999,2000
  *
- * $Revision: 1.53.2.1 $
+ * $Revision: 1.53.2.2 $
  */
 
 #include <linux/config.h>
@@ -1070,7 +1070,7 @@
 	cqr->device = device;
 	cqr->expires = 5 * 60 * HZ;	/* 5 minutes */
 	cqr->lpm = LPM_ANYPATH;
-	cqr->retries = 2;
+	cqr->retries = 256;
 	cqr->buildclk = get_clock();
 	cqr->status = DASD_CQR_FILLED;
 	return cqr;

